src/Toml/Printer.hs

```diff
@@ -134,36 +127,54 @@ prettyKeyValue options i =
             . formatTime defaultTimeLocale "%z"
 
 -- | Returns pretty formatted tables section of the 'TOML'.
-prettyTables :: PrintOptions -> Int -> Text -> PrefixMap TOML -> Text
-prettyTables options i pref =
-    Text.concat . map (prettyTable . snd) . orderWith options . HashMap.toList
+prettyTables :: PrintOptions -> Int -> Text -> PrefixMap TOML -> [Text]
+prettyTables options i pref = mapOrdered (prettyTable . snd) options
   where
-    prettyTable :: PrefixTree TOML -> Text
+    prettyTable :: PrefixTree TOML -> [Text]
     prettyTable (Leaf k toml) =
-        let name = getPref k in mconcat
-            [ tabWith options i
-            , prettyTableName name
-            , prettyTomlInd options (succ i) name toml ]
+        let name = addPrefix k pref
+        in "": tabWith options i <> prettyTableName name :
+             dropWhile (== "") (prettyTomlInd options (succ i) name toml)
+
     prettyTable (Branch k mToml prefMap) =
-        let name  = getPref k
+        let name  = addPrefix k pref
             nextI = succ i
             toml  = case mToml of
-                        Nothing -> ""
+                        Nothing -> []
                         Just t  -> prettyTomlInd options nextI name t
-        in mconcat
-            [ tabWith options i
-            , prettyTableName name
-            , toml
-            , prettyTables options nextI name prefMap ]
-
-    -- Adds next part of the table name to the accumulator.
-    getPref :: Key -> Text
-    getPref k = case pref of
-        "" -> prettyKey k
-        _  -> pref <> "." <> prettyKey k
+        in "": tabWith options i <> prettyTableName name :
+             dropWhile (== "") (toml ++ prettyTables options nextI name prefMap)
```

Comment:
Yeah, I don't think it will be easy to understand what's going on here in future

Reply:
It's not too hard to explain: a `""` will print as an empty line, we always want an empty line before a table/array name, but never after.

Reply:
In that case adding some comment to this piece of code might be useful 👍